### ISSUE TITLE
Two Factor Authentication

### ISSUE THREAD

**Issue by ganaraj-savant** [passepartoutvpn/tunnelkit#269]
Is there any way to achieve? I'm getting a challenge from the server on initial authentication, after which I need to show a screen to get the OTP from user. But how to keep the tunnel in waiting/deferred state and also update the state to UI layer where I can ask the user for OTP?

Thanks in advance!

**Comment by ganaraj-savant** [comment]
Hi Keeshux,

I have kinda got over it. I'm able to get the OTP from the user and pass the updated credentials to the tunnel provider. Now the only challenge is, how do I update it to the server so that it re-authenticates and a connection is established?

Thanks!

**Comment by keeshux** [comment]
Is this something that you are able to accomplish somehow with OpenVPN Connect?

**Comment by ganaraj-savant** [comment]
> Is this something that you are able to accomplish somehow with OpenVPN Connect?

Yes, I was able achieve 2FA with TunnelKit. I got the solution from Apple Developer Forums, find the thread link here: https://developer.apple.com/forums/thread/712046

Then I started the tunnel again for re-authentication with the OTP.

**Comment by keeshux** [comment]
@ganaraj-savant I see you mention AUTH_FAILED, which sounds like OpenVPN. How do you tell a plain authentication failure from a 2FA challenge request?
